### ISSUE TITLE
Feature Request: autodiscoovery for ASDF support

### ISSUE THREAD

**Issue by v1v** [updatecli/updatecli#2274]
### Is your feature request related to a problem?

[ASDF](https://asdf-vm.com/manage/configuration.html#tool-versions) uses `.tool-versions` to define tools and versions in the repository.

https://github.com/updatecli/updatecli/pull/2265 trying to add some support using the `toolversion` kind, but it's not implementing the autodiscovery.

There is a discussion about supporting it, and we have agreed to discuss it in an Issue before anything else is implemented.

### Solution you'd like

Support autodiscovery for the well-defined tools used in `.tool-versions`, such as `golang`, `python`, `java`.

### Anything else?

It's something to consider, though. It's optional for the first part of this conversation, but it could be a follow-up if this is something agreed to be pursued.

How do we support other tools defined in the `.tool-versions` file using a plugin-based approach? Or something that can help avoid ad-hoc implementations.

For instance, if I use a tool called `my-super-tool`, its version is `0.0.1`, and already defined in `.tool-versions`. How do I configure updatecli to bump the version with `autodiscovery` if `my-super-tool` follows the GitHub releases in a GitHub release project called `my-super-tool`? What if the GitHub repository name is different? And so on :)

Using a plugin-based approach, how do we support other tools defined in the `.tool-versions` file

**Comment by olblak** [comment]
> Support autodiscovery for the well-defined tools used in .tool-versions, such as golang, python, java.

I am wondering if we can find an API that provide those information, otherwise we'll need to implement specific updatecli plugins for all of them which I have concern of work needed.

For example: we already have one for Jenkins, ruby would be simple as we can retrieve information from https://raw.githubusercontent.com/ruby/www.ruby-[lang.org/master/_data/downloads.yml](http://lang.org/master/_data/downloads.yml)

I am aware of pkgs.org but it requires an access token.
I once find another website but I don't remember the name

**Comment by dduportal** [comment]
> I am wondering if we can find an API that provide those information, otherwise we'll need to implement specific updatecli plugins for all of them which I have concern of work needed.

Yes, however I'm sure we'll find some common grounds: the ruby, golang or temurin ASDF plugins for instance. But that will be quite some work if we go native.

@v1v do you think it is acceptable to have your resource running the expected `asdf <...>` commands inside a controller environment (using custom `PATH` in a tmpdir)? But that would restrict execution on Bash only though (installing asdf + plugins in a tmp dir is easy though, even without touching the current env.)

**Comment by v1v** [comment]
I understand the threshold but I'd prefer if `asdf` is not needed, otherwise, what `asdf` version is the one to be used and other questions around dependency management. 

> I am wondering if we can find an API that provide those information

I've been thinking about whether the technical implementation for the updatecli policies that use the github packages could be adapted to support the API contract with the list of tools/versions.

**Comment by olblak** [comment]
I am wondering if we can build a parser for [repology.org](https://repology.org)
For example https://repology.org/project/python/versions contains a lot of versions

**Comment by dduportal** [comment]
> I am wondering if we can build a parser for [repology.org](https://repology.org) For example https://repology.org/project/python/versions contains a lot of versions

Good idea! However it us unrelated to `asdf` (or did I miss something?)

**Comment by olblak** [comment]
This is the part I wonder, I am not familiar enough with asdf to know which version should be needed in the configuration file. Would it be the "upstream" one? or the one available by the local distribution?

**Comment by dduportal** [comment]
> This is the part I wonder, I am not familiar enough with asdf to know which version should be needed in the configuration file. Would it be the "upstream" one? or the one available by the local distribution?

That is my point: alas it is a per-plugin basis. And since the native language of the plugins is `bash`, then you have to, either:

- Call `asdf` commands
  - Pros: native `asdf` behavior, no need to re-implement in `updatecli`
  - Cons: requirement on the `asdf` + `bash` tools for `updateci` (as pointed out by @v1v )

- Do an ASDF plugin by plugin analysis and implement the same logic in `updatecli` native. E.g. for each plugin: "if it is using upstream then use resue it. If not then deep dive the ASDF plugin and see what it does".
  -  Pros: can run without requiring `shell` neither `asdf`
  - Cons: maintenance effort (example for the widely used ASDF java plugin: it integrates "numerous" Java upstreams: Zulu, Temurin, old OpenJDK, etc.) as we need case by case

- Use shell commands (without ASDF) to retrieve the plugin shell code and call the shell functions (usually called by `asdf`) to get output, parse it and use it
  - Pros: intermediate: no requirement on `asdf`, can run in Cygwin, Linux, Unix, even powershell if `bash` is somewhere to be found (git SCM on widnows has binaries). Always up to date with plugins as it follows the same interface as ASDF plugins.
  - Cons: still requires shell

**Comment by olblak** [comment]
Without knowing the effort needed to implement each plugin I would prefer to go down that approach, I am not a big fan of running external commands from Updatecli. We do it for some autodiscovery plugins but I would rather consider this as the exception.`
